src/main.rs

```diff
@@ -135,7 +138,13 @@ async fn provision() -> Result<(), anyhow::Error> {
     // immediately in a single failure, instead it should fall back to the other
     // mechanism. So it is not a good idea to use `?` for query() or
     // get_environment().
-    let instance_metadata = imds::query(&client).await.ok();
+    let instance_metadata = util::retry_loop(
```

Comment:
While having a `util` module isn't unprecedented, it does mean library users have to know about it and that they want to use it in conjunction with `imds::query`. Given the concrete retry policy in #63 I wonder if we even want to expose any tune-ables to users.

The API could either stay `imds::query(&client)` and we handle the retries and timeout internally or, if we want to expose options to the user, change it to be `imds::query(&query, retry_interval, total_timeout)`.

I think what it could look like internal to `imds::query` would be something like this (haven't compile-tested, not formatted for tidiness):
```Rust
// Issue #63 says 5 minutes for IMDS
let response = tokio::time::timeout(Duration::from_secs(5 * 60), async {
    let interval = Duration::from_secs(2);

    loop {
        if let Ok(response) = client.get(url).timeout(Duration::from_secs(30).send().await {
            // Do some response inspection here for what to retry, but basic version is below
            if response.status() == 200 {
                break response;
            }
        }
        // not sure if backoff of 1s described in #63 means we should do `interval += Duration::from_secs(1)`, flat 2s
        // sleep here
        tokio::time::sleep(interval.clone()).await;
    }
    
}).await?;
```

Basically, write a future that loops forever with the retry, and then enforce the overall timeout with tokio. That way we don't need to think about intervals we sleep, time we spent on individual request timeouts, etc.

Reply:
Done, thanks.